### PR TITLE
Fix Consul Service PreKill Hook

### PR DESCRIPTION
**Pull request by schmichael**
...and port a bunch of tests from 0.8. All passing with `-race` enabled.

### CODE REVIEW

File: client/allocrunner/taskrunner/service_hook.go [hunk 1]
```diff
@@ -140,7 +140,7 @@ func (h *serviceHook) Update(ctx context.Context, req *interfaces.TaskUpdateRequ
 	return h.consul.UpdateTask(oldTaskServices, newTaskServices)
 }
 
-func (h *serviceHook) Killing(ctx context.Context, req *interfaces.TaskPreKillRequest, resp *interfaces.TaskPreKillResponse) error {
+func (h *serviceHook) PreKilling(ctx context.Context, req *interfaces.TaskPreKillRequest, resp *interfaces.TaskPreKillResponse) error {
```
Comment (notnoop) [inline client/allocrunner/taskrunner/service_hook.go]: What's the cause of this change?  Was it a missed regression?

In the future, would be nice to have a better way to detect hook interface mismatch with type system or other means.

File: client/allocrunner/taskrunner/task_runner_test.go [hunk 2]
```diff
@@ -767,6 +770,247 @@ func TestTaskRunner_CheckWatcher_Restart(t *testing.T) {
...
+	require.Equal(t, structs.TaskStatePending, tr.TaskState().State)
+
+	// Unblock vault token
+	close(waitCh)
+
+	// TR should exit now that it's unblocked by vault as its a batch job
+	// with 0 sleeping.
+	select {
+	case <-tr.WaitCh():
+	case <-time.After(time.Duration(testutil.TestMultiplier()*15) * time.Second):
```
Comment (notnoop) [inline client/allocrunner/taskrunner/task_runner_test.go]: I understand that it's the pattern elsewhere in file, but I find this somewhat hard to parse.  I find the following easier to read

```suggestion
	case <-time.After(15 * time.Second * time.Duration(testutil.TestMultiplier())):
```

File: client/allocrunner/taskrunner/task_runner_test.go [hunk 3]
```diff
@@ -767,6 +770,247 @@ func TestTaskRunner_CheckWatcher_Restart(t *testing.T) {
...
+	// Check that the token is on disk
+	tokenPath := filepath.Join(conf.TaskDir.SecretsDir, vaultTokenFile)
+	data, err := ioutil.ReadFile(tokenPath)
+	require.NoError(t, err)
+	require.Equal(t, token, string(data))
+
+	// Check the token was revoked
+	testutil.WaitForResult(func() (bool, error) {
+		if len(vaultClient.StoppedTokens()) != 1 {
+			return false, fmt.Errorf("Expected a stopped token: %v", vaultClient.StoppedTokens())
```
Comment (notnoop) [inline client/allocrunner/taskrunner/task_runner_test.go]: nit pick : When `StoppedTokens()` is empty, the error message isn't useful:
```suggestion
			return false, fmt.Errorf("Expected a stopped token %q but found: %v", token, vaultClient.StoppedTokens())
```

File: client/vaultclient/vaultclient_testing.go [hunk 4]
```diff
@@ -1,45 +1,53 @@
...
+	"sync"
+
 	"github.com/hashicorp/nomad/helper/uuid"
 	"github.com/hashicorp/nomad/nomad/structs"
 	vaultapi "github.com/hashicorp/vault/api"
 )
 
-// MockVaultClient is used for testing the vaultclient integration
+// MockVaultClient is used for testing the vaultclient integration and is safe
+// for concurrent access.
```
Comment (notnoop) [inline client/vaultclient/vaultclient_testing.go]: Good catch - the TR tests do run with multiple goroutines!

File: client/allocrunner/taskrunner/service_hook.go (hunk 1 above)
Comment (schmichael) [reply client/allocrunner/taskrunner/service_hook.go]: The method on the interface was changed in #5207 and this implementation was missed. We do static assertions in tests, but this hook hadn't implemented those yet: https://github.com/hashicorp/nomad/pull/5319/files#diff-5cc9be3181b1e0e392ff275b17fbdfe1R12

Any other ideas on how to avoid mistakes like this are welcome, but we should always have some sort of integration test (like the one I added that discovered this bug) to catch it.

Comment (notnoop) [reply client/allocrunner/taskrunner/service_hook.go]: good point.  Static checks are quite useful as a cheap baseline.  Would be useful to do a quick audit of all hooks as well.  Thanks for catching it!
